libraries/message/src/config.rs

```diff
@@ -1,129 +1,87 @@
-    convert::Infallible,
-    fmt,
-    str::FromStr,
     time::Duration,
 };
 
-#[derive(
-    Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, JsonSchema,
-)]
-pub struct NodeId(String);
-
-impl FromStr for NodeId {
-    type Err = Infallible;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self(s.to_owned()))
-    }
-}
-
-impl From<String> for NodeId {
-    fn from(id: String) -> Self {
-        Self(id)
-    }
-}
-
-impl std::fmt::Display for NodeId {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        std::fmt::Display::fmt(&self.0, f)
-    }
-}
-
-impl AsRef<str> for NodeId {
-    fn as_ref(&self) -> &str {
-        &self.0
-    }
-}
-
-#[derive(
-    Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, JsonSchema,
-)]
-pub struct OperatorId(String);
-
-impl FromStr for OperatorId {
-    type Err = Infallible;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self(s.to_owned()))
-    }
-}
-
-impl From<String> for OperatorId {
-    fn from(id: String) -> Self {
-        Self(id)
-    }
-}
-
-impl std::fmt::Display for OperatorId {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        std::fmt::Display::fmt(&self.0, f)
-    }
-}
-
-impl AsRef<str> for OperatorId {
-    fn as_ref(&self) -> &str {
-        &self.0
-    }
-}
-
-#[derive(
-    Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, JsonSchema,
-)]
-pub struct DataId(String);
-
-impl From<DataId> for String {
-    fn from(id: DataId) -> Self {
-        id.0
-    }
-}
-
-impl From<String> for DataId {
-    fn from(id: String) -> Self {
-        Self(id)
-    }
-}
-
-impl std::fmt::Display for DataId {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        std::fmt::Display::fmt(&self.0, f)
-    }
-}
+use once_cell::sync::OnceCell;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
 
-impl std::ops::Deref for DataId {
-    type Target = String;
+pub use crate::id::{DataId, NodeId, OperatorId};
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+pub struct NodeRunConfig {
+    /// Inputs for the nodes as a map from input ID to `node_id/output_id`.
+    ///
+    /// e.g.
+    ///
+    /// inputs:
+    ///
+    ///   example_input: example_node/example_output1
+    ///
+    #[serde(default)]
+    pub inputs: BTreeMap<DataId, Input>,
+    /// List of output IDs.
+    ///
+    /// e.g.
+    ///
+    /// outputs:
+    ///
+    ///  - output_1
+    ///
+    ///  - output_2
+    #[serde(default)]
+    pub outputs: BTreeSet<DataId>,
 }
 
-impl AsRef<String> for DataId {
-    fn as_ref(&self) -> &String {
-        &self.0
-    }
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields, from = "InputDef", into = "InputDef")]
```

Comment:
I guess we should probably not `deny_unknown_fields` for the sake of forward compatibility. If its part of the message library that might not get updated regularly.

Reply:
Good point, removed.